speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp

```diff
@@ -35,31 +81,327 @@ inline std::string GetTimeAsString(time_t p_date)
+void TestDataStorageSave(bool p_storeEnvironment, bool p_storeCar, bool p_storeControls, bool p_storeDecisions, bool p_storeMeta)
+{
+    Random random;
+    tDataToStore params = {
+        p_storeEnvironment,
+        p_storeCar,
+        p_storeControls,
+        p_storeDecisions,
+        p_storeMeta};
+    FileDataStorage fileDataStorage;
+
+    // Create a string to save all intended random data to
+    std::stringstream expected;
+
+    // Initialise buffer file
+    GET_DUMMY_TIMES;
+    std::experimental::filesystem::path path = fileDataStorage.Initialize(params, DUMMY_INITIALISATION_PARAMETERS);
+    expected << DUMMY_INITIALISATION_FILE_ENTRIES;
+    if (p_storeEnvironment)
+    {
+        expected << "GameState\n";
+    }
+    if (p_storeControls)
+    {
+        expected << "UserInput\n";
+    }
+
+    for (int i = 0; i < 3; i++)
+    {
+        // Save with random data
+        GET_RANDOM_TICKCOUNT;
+        GET_RANDOM_CAR;
+        GET_RANDOM_SITUATION;
+        fileDataStorage.Save(&car, &situation, tickCount);
+
+        // Define our expectations
+        expected << std::to_string(tickCount) << "\n";
+        if (p_storeEnvironment)
+        {
+            WRITE_EXPECTED_ENVIRONMENT;
+        }
+        if (p_storeControls)
+        {
+            WRITE_EXPECTED_CONTROLS;
+        }
+    }
+
+    // Finish the buffer file
+    expected << "END";
+    fileDataStorage.Shutdown();
+
+    // Read the written file
+    std::ifstream reader(path);
+    std::string fileContents((std::istreambuf_iterator<char>(reader)), std::istreambuf_iterator<char>());
+    reader.close();
+
+    // Check contents
+    ASSERT_STREQ(fileContents.c_str(), expected.str().c_str());
+}
+
+/// @brief Run the @link #TestDataStorageSave(bool,bool,bool,bool,bool) test with all possible pairs of datasets enabled at least once.
+TEST(FileDataStorageTests, TestDataStorageSingle)
+{
+    bool booleans[2]{true, false};
+    PairWiseTest(TestDataStorageSave, booleans, 2, booleans, 2, booleans, 2, booleans, 2, booleans, 2);
+}
+
+void TestDataStorageSaveDecisions(bool p_storeDecisions, bool p_doSteer, bool p_doBrake, bool p_doAccel, bool p_doGear, bool p_doLights)
+{
+    Random random;
+    tDataToStore params = {
+        false,
+        false,
+        false,
+        p_storeDecisions,
+        false};
+    FileDataStorage fileDataStorage;
+
+    // Create a string to save all intended random data to
+    std::stringstream expected;
+
+    // Initialise buffer file
+    GET_DUMMY_TIMES;
+    std::experimental::filesystem::path path = fileDataStorage.Initialize(params, DUMMY_INITIALISATION_PARAMETERS);
+    expected << DUMMY_INITIALISATION_FILE_ENTRIES;
+
+    DecisionTuple decisions = GenerateDecisions(random, p_doSteer, p_doBrake, p_doAccel, p_doGear, p_doLights);
+
+    // Save any tick
+    GET_RANDOM_TICKCOUNT;
+    fileDataStorage.Save(nullptr, nullptr, tickCount);
+    expected << std::to_string(tickCount) << "\n";
```

Comment:
Change all '\n' characters with std::endl as enters are different on windows and linux